Refactor module resolution `Extensions`, fix lookup priorities

andrewbranch opened the pull request:
<!--
Thank you for submitting a pull request!

Please verify that:
* [ ] There is an associated issue in the `Backlog` milestone (**required**)
* [ ] Code is up-to-date with the `main` branch
* [ ] You've successfully run `hereby runtests` locally
* [ ] There are new or updated unit tests validating the change

Refer to CONTRIBUTING.MD for more details.
  https://github.com/Microsoft/TypeScript/blob/main/CONTRIBUTING.md

** Please don't send typo fixes! **
Please don't send a PR solely for the purpose of fixing a typo, unless that
typo truly hurts understanding of the text. Each PR represents work for the
maintainers, and that work should provide commensurate value.

If you're interested in sending a PR, the issue tracker has many issues marked `help wanted`.
-->

Fixes two forever-standing resolution bugs:

1. The order of adding extensions vs. replacing extensions was backwards. Previously, for an import of `"./foo.js"`, we would look up `"./foo.js.ts"` at a higher priority than `"./foo.ts"`. This has been fixed in every module resolution mode.
2. Each module resolution algorithm previously ran two passes: one looking for `.ts` or `.d.ts` files, and then another full pass looking for `.js` files if the former returned nothing. This results in some priorities that are obviously wrong under `allowJs`, and since `allowJs` doesn’t actually affect resolution results (only whether JS resolutions are errors after the fact), probably always wrong by extension. An example:
   ```ts
   // dir/index.ts
   export {}
   // dir.js
   export {}
   // main.ts
   import "./dir"
   ```
   Here, `import "./dir"` should resolve to `dir.js`, but previously resolved to `dir/index.ts` in every module resolution algorithm. This has been fixed only in `node16` and `nodenext` in an attempt to limit breaking changes to `classic` and `node`. Those algorithms now try all valid extensions in priority order for each lookup location. The exception is that for a given `node_modules` folder searched when looking up a non-relative name, the implementation package is searched for `.ts`/`.d.ts` files only, then the `@types` package is searched, then the implementation package is searched again for `.js` files. Making this change required a pretty extensive refactor of how the algorithms deal with extensions, and I think the result is much easier to reason about.

For file system hits and watches, this should be a wash—depending on what files are where, a given resolution might have a few more lookups, a few less, or the same number but in a different order.

andrewbranch commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -82,15 +82,23 @@ interface PathAndExtension {
-    DtsOnly,    /** Only '.d.ts' */
-    TsOnly,     /** '.[cm]tsx?' but not .d.ts variants */
+const enum Extensions {
+    TypeScript  = 1 << 0, // '.ts', '.tsx', '.mts', '.cts'
+    JavaScript  = 1 << 1, // '.js', '.jsx', '.mjs', '.cjs'
+    Declaration = 1 << 2, // '.d.ts', etc.
+    Json        = 1 << 3, // '.json'
+
+    ImplementationFiles = TypeScript | JavaScript,
+}
```
`Extensions` is bit flags now, and `TSConfig` is changed to be an `isConfigLookup` on `ModuleResolutionState`.

andrewbranch commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -1575,28 +1604,26 @@ function loadModuleFromFileNoPackageId(extensions: Extensions, candidate: string
         }
     }
-
-    return loadModuleFromFileNoImplicitExtensions(extensions, candidate, onlyRecordFailures, state);
```
This is the extension adding/replacing order swap

andrewbranch commented on tests/cases/compiler/moduleResolutionWithExtensions_preferTs.ts:
```diff
@@ -1,10 +0,0 @@
-export default 0;
-
-// @Filename: /a.ts
-import b from "./b";
```
This test asserted the incorrect behavior. It’s replaced by the two new tests.

jakebailey commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -82,15 +82,23 @@ interface PathAndExtension {
+    ImplementationFiles = TypeScript | JavaScript,
+}
+
+function formatExtensions(extensions: Extensions) {
```
FWIW you can get a pretty similar result with `Debug.formatEnum`. Post-modules that's something I want to yank out into a module for use outside of Debug.

andrewbranch replied:
I didn’t use it mostly because it was in `Debug` and this production code... I know it would work without any issue but it felt wrong.

jakebailey replied:
Yeah, I wouldn't have mentioned it but this seemed like it was just in tracing which straddles the line a little.

jakebailey reviewed (approved):
LGTM but I'll leave it for Wesley to look at tomorrow.

andrewbranch commented:
> Those algorithms now try all valid extensions in priority order for each lookup location. The exception is that for a given `node_modules` folder searched when looking up a non-relative name, the implementation package is searched for `.ts`/`.d.ts` files only, then the `@types` package is searched, then the implementation package is searched again for `.js` files.

I think this might be a problem for monorepos where `@types` packages might be hoisted to the monorepo root level while implementation packages might get installed in the monorepo subpackages due to different version requirements.

sheetalkamat commented:
> this has been fixed only in node16 and nodenext in an attempt to limit breaking changes to classic and node.

if this is breaking change anyway why not correct it for all algorithms.. Havent looked at changes yet and will do soon but does it simplify code even further or it makes it worse. Apart from being breaking change does it have any other reasoning not to break these two modes. Will it not do right thing.

andrewbranch commented:
I intend for `classic` and `node` to be essentially unmaintained after 5.0, so I don’t see a great reason to “fix” them if nobody has complained about this, but `node16` and `nodenext` are new enough to bother fixing. It’s essentially a one-line change to opt `node` and `classic` into or out of the fix, so either way has little impact on code complexity.

andrewbranch commented:
I really expected a couple monorepos to fail there. Interesting.

andrewbranch commented:
Ah, the lack of failures could be due to low usage of `node16` and `nodenext`. I’ll try applying the same fix to `node` and see how much stuff changes.

andrewbranch commented:
@typescript-bot test top100

sheetalkamat commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -2447,11 +2428,23 @@ function loadModuleFromImmediateNodeModulesDirectory(extensions: Extensions, mod
+    //   1. TS/DTS files in the implementation package
+    //   2. DTS files in the @types package
+    //   3. JS files in the implementation package
+    const priorityExtensions = extensions & (Extensions.TypeScript | Extensions.Declaration);
+    const secondaryExtensions = extensions & ~(Extensions.TypeScript | Extensions.Declaration);
```
The primary and secondary names here with the comments is confusing since they correspond to points 1 and 3
I kept thinking how is secondary same as option 2 and then when i saw usage it became clear that its actually point 3,

sheetalkamat commented on src/compiler/moduleNameResolver.ts:
```diff
@@ -1340,45 +1350,51 @@ function nodeNextModuleNameResolver(moduleName: string, containingFile: string,
-        extensions = [Extensions.TsOnly];
-        if (compilerOptions.allowJs) extensions.push(Extensions.JavaScript);
-        if (compilerOptions.resolveJsonModule) extensions.push(Extensions.Json);
+        extensions = Extensions.ImplementationFiles;
```
this is the part where we are saying always looks for js and report error if allowJs is not on correct.

andrewbranch replied:
This is actually only used by Find Source Definition, which always passes `allowJs`. But I removed the condition to be consistent with other resolution modes which always resolve JS, yes.

andrewbranch replied:
And to be clear, normal resolution has resolved JS and reported errors afterwards for like 6 years; that’s not new here 😄

andrewbranch commented:
@typescript-bot perf test this

andrewbranch commented:
@typescript-bot test top100 again, even though you haven’t finished the last one

andrewbranch commented:
^ Yeah, that’s what I expected. I’ve reverted the node_modules searching algorithm back to two full passes, so I expect the second top100 run to come back clean. Updated the PR description:

> The exception is that node_modules lookups still happen in two passes: one that searches the implementation package followed by the `@types` package looking for `.ts`/`.d.ts` files for each `node_modules` folder found up the directory spine, followed (if necessary) by another pass up the directory spine looking for `.js` files in implementation packages.
